eegnb/experiments/Experiment.py

```diff
@@ -0,0 +1,133 @@
+""" 
+Initial run of the Experiment Class Refactor base class
+
+Specific experiments are implemented as sub classes that overload a load_stimulus and present_stimulus method
+
+
+Running each experiment:
+obj = VisualP300({parametrs})
+obj.present()
+"""
+
+from abc import ABC, abstractmethod
+from psychopy import prefs
+#change the pref libraty to PTB and set the latency mode to high precision
+prefs.hardware['audioLib'] = 'PTB'
+prefs.hardware['audioLatencyMode'] = 3
+
+import os
+from time import time
+from glob import glob
+from random import choice
+from optparse import OptionParser
+import random
+
+import numpy as np
+from pandas import DataFrame
+from psychopy import visual, core, event
+
+from eegnb import generate_save_fn
+from eegnb.devices.eeg import EEG
+
+class Experiment:
+
+    def __init__(self, exp_name, duration, eeg, save_fn, n_trials, iti, soa, jitter):
+        """ Anything that must be passed as a minimum for the experiment should be initialized here """
+
+        self.exp_name = exp_name
+        self.instruction_text = """\nWelcome to the {} experiment!\nStay still, focus on the centre of the screen, and try not to blink. \nThis block will run for %s seconds.\n
+        Press spacebar to continue. \n""".format(self.exp_name)
+        self.duration = duration
+        self.eeg = eeg
+        self.save_fn = save_fn
+        self.n_trials = n_trials
+        self.iti = iti
+        self.soa = soa
+        self.jitter = jitter
+        
+    @abstractmethod
+    def load_stimulus(self):
+        """ Needs to be overwritten by specific experiment """
+        raise NotImplementedError
+
+    @abstractmethod
+    def present_stimulus(self):
+        raise NotImplementedError
+
+    def setup(self):
+
+        self.record_duration = np.float32(self.duration)
+        self.markernames = [1, 2]
+        
+        # Setup Trial list -> Common in most (csv in Unicorn)
+        self.parameter = np.random.binomial(1, 0.5, self.n_trials)
+        self.trials = DataFrame(dict(parameter=self.parameter, timestamp=np.zeros(self.n_trials)))
+
+        # Setup Graphics 
+        self.mywin = visual.Window([1600, 900], monitor="testMonitor", units="deg", fullscr=True) 
+        
+        # Needs to be overwritten by specific experiment
+        self.stim = self.load_stimulus()
+        
+        # Show Instruction Screen
+        self.show_instructions()
+
+        if self.eeg:
+            if save_fn is None:  # If no save_fn passed, generate a new unnamed save file
+                random_id = random.randint(1000,10000)
+                self.save_fn = generate_save_fn(self.eeg.device_name, "visual_n170", random_id, random_id, "unnamed")
+                print(
+                    f"No path for a save file was passed to the experiment. Saving data to {self.save_fn}"
+                )
+    
+    def present(self):
+        """ Do the present operation for a bunch of experiments """
+
+        self.setup()
```

Comment:
As discussed, let's add an argument here so that instructions can optionally be skipped. Default behaviour should be to show however.